### PR TITLE
Port upstream libffi 3.4.4 for x86_64

### PR DESCRIPTION
**Pull request by andreittr**
This change ports the parts of libffi 3.4.4 relevant to x86 to unikraft.

Part of work on Python 3.10 (https://github.com/unikraft/lib-python3/pull/15).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,8 @@
-# lib-libffi
- A portable foreign-function interface library
+libffi for Unikraft
+========================
+
+This is a port of [libffi](https://sourceware.org/libffi/) to Unikraft.
+
+Please refer to the `README.md` as well as the documentation in the
+`doc/` subdirectory of the main unikraft repository for further
+information.
```
Comment (StefanJum) [inline README.md]:
```suggestion
Please refer to the `README.md` as well as the documentation in the `doc/` subdirectory of the main unikraft repository for further information.
```
Use one sentence per line and one line per sentence.

File: README.md [hunk 2]
```diff
@@ -1,2 +1,8 @@
-# lib-libffi
- A portable foreign-function interface library
+libffi for Unikraft
+========================
```
Comment (StefanJum) [inline README.md]:
```suggestion
# `libffi` for Unikraft
```
Use md format, not rst.

**Review by StefanJum — COMMENTED** [review]
@andreittr Will we only support this on `x86_64` for this release?

**Comment by andreittr** [comment]
> @andreittr Will we only support this on `x86_64` for this release?

Only the x86_64 work is done thus far (and getting arm done before release is unlikely); perhaps @razvand can opine whether to upstream this in current release or postpone until arm work is also done.

**Comment by andreittr** [comment]
Update: made changes as discussed in https://github.com/unikraft/lib-python3/pull/15.

**Review by mariasfiraiala — CHANGES_REQUESTED** [review]
Thanks @andreittr, I've tested this together with unikraft/lib-python3#15 and it builds just fine.

I am, however, reluctant to have a config option that _disables_ something. I am aware of the usecase, but it still weirds me out; it's kinda counterintuitive to select an option in order to _not_ have the definitions..

Anyway, it's a personal preference, I'll approve the changes once the app-python3 PR is out in order to have a run too.

**Comment by andreittr** [comment]
> Thanks @andreittr, I've tested this together with [unikraft/lib-python3#15](https://github.com/unikraft/lib-python3/pull/15) and it builds just fine.
> 
> I am, however, reluctant to have a config option that _disables_ something. I am aware of the usecase, but it still weirds me out; it's kinda counterintuitive to select an option in order to _not_ have the definitions..
> 
> Anyway, it's a personal preference, I'll approve the changes once the app-python3 PR is out in order to have a run too.

Oh I completely agree with you on this front, negative options are always more confusing than positive ones. Pity that Kconfig does not support forcefully unselecting a value. Unless and until that becomes a thing I'm afraid we're stuck with this inelegant solution. Other alternatives off the top of my head would be

(a) Leave things as before (positive option, py3 leaves it alone); upside: no negative config options; downside: if option is default 'y' -> py3 won't build unless tweaked, if option defaults to 'n' -> libffi won't build by itself unless tweaked; either way requires documentation and/or tribal knowledge

(b) Leave things as before but check & error out in the Makefiles; upside: no negative config options; downside: we leave to manual trial-and-error something that can and should be automated, and no amount of explaining Kconfig's limitations will lessen users' frustration

(c) Keep negative option but make it invisible; upside: no user confusion by default, libs can still opt out of `ffi_type_*` definitions when needed; downside: blackmagic™ happends during configuration, it's not clear where to disable `ffi_type_*` stuff if needed for e.g. a custom lib/app

(d) hack our own fork of Kconfig that can do "select CONFIGVAL=n" (I'm ~90% joking with this one, totally not volunteering to do smth like this)

Thoughts?

**Comment by mariasfiraiala** [comment]
Now that you've laid out the options we have, @andreittr, I'd go with the black magik stuff (October it's rapidly approaching :mage_woman:). It's the one that has the least damaging downsides, imo. @StefanJum opinions on this?

Btw, really really nice presentation of the alternatives @andreittr :100:

**Comment by StefanJum** [comment]
> Oh I completely agree with you on this front, negative options are always more confusing than positive ones. Pity that Kconfig does not support forcefully unselecting a value. Unless and until that becomes a thing I'm afraid we're stuck with this inelegant solution.

Right on point. I would honestly keep the negative option as it is now. It is ugly, but at least it's clear and things work out of the box.
From the options you provided I think we could go with (c) (so everything will still work out of the box), but again, I think the current way is cleaner, as ugly as it is.

**Comment by mariasfiraiala** [comment]
We shall have it as it is then, all good!

**Comment by andreittr** [comment]
> I'd go with the black magik stuff

>  I think we could go with (c) (so everything will still work out of the box), but again, I think the current way is cleaner, as ugly as it is.

Just thought of a decent compromise between this PR and option (c): keep the option visible but rename its user prompt to a positive statement like "Allow other libraries to supply `ffi_type_*` definitions". Best of both worlds?

**Comment by StefanJum** [comment]
> Just thought of a decent compromise between this PR and option (c): keep the option visible but rename its user prompt to a positive statement like "Allow other libraries to supply `ffi_type_*` definitions". Best of both worlds?

Nice, I think we can do something like `Use external ffi_type_* definitions` and also provide a `help` message that will state that if the option is selected, the `ffi_type_*` are not defined, and are to be supplied by an external library.

**Review by RaduNichita — COMMENTED** [review]
I also tested this as part of python3.10 PR series and it builds and runs fine. Thanks, @andreittr  for this nice work! :medal_sports: 

I would also go for option c, I think it's the most appropriate for the reasons that @StefanJum  and @mariasfiraiala have already mentioned.

**Review by StefanJum — APPROVED** [review]
Thank you @andreittr :rocket: 
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

**Review by mariasfiraiala — APPROVED** [review]
Thanks for this long journey!

Reviewed-by: Maria Sfiraiala <maria.sfiraiala@gmail.com>

**Review by RaduNichita — APPROVED** [review]
Let's merge this. Thanks @andreittr for the contribution! :top: 

Reviewed-by: Radu Nichita <radunichita99@gmail.com>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**